File: Exp1/Sudoku-Python/Sudoku.py
import copy
import re


class Grid:
    """
    存放解析后的数独字符串的二维矩阵以及相关的访问矩阵行列的方法
    """
    GRID_SIZE = 9
    BOX_SIZE = 3

    def __init__(self):
        self.grid = [[0 for j in range(Grid.GRID_SIZE)] for i in range(Grid.GRID_SIZE)]

    def getRow(self, row)->list:
        """获取行"""
        return self.grid[row]

    def getCol(self, col)->list:
        """获取列"""
        return [each[col] for each in self.grid]

    def getBox(self, row, col)->list:
        """获取(row,col)所在的3x3宫格"""
        res = []
        rowStart, colStart = (row // 3) * 3, (col // 3) * 3
        for i in range(rowStart, rowStart + 3):
            for j in range(colStart, colStart + 3):
                res.append(self.grid[i][j])
        return res


class Sudoku(Grid):
    def __init__(self, sudokuStr):
        super(Sudoku, self).__init__()
        self.sudokuStr = sudokuStr

    def checkStr(self)->bool:
        """
        对输入的数独字符串进行格式检查
        :return:
        """
        if not self.sudokuStr:
            return False
        # 检查是否只包含数字1-9和字符'0'
        if not re.match(r'^[0-9]{81}$', self.sudokuStr):
            return False

        return True

    def parseStr(self)->bool:
        """
        将输入的数独字符串解析为二维矩阵
        :return:
        """
        # 格式检查
        if not self.checkStr():
            return False

        # str转矩阵
        for row in range(Grid.GRID_SIZE):
            for col in range(Grid.GRID_SIZE):
                self.grid[row][col] = int(self.sudokuStr[row * Grid.GRID_SIZE + col])
        return True

    def isValid(self, row, col, val) -> bool:
        """
        根据val检查所在行、列和宫格Box是否满足数独的规则要求
        :param row: 行号
        :param col: 列号
        :param val: 需要检查的值
        :return:
        """

        # 检查行
        if val in self.getRow(row):
            return False

        # 检查列
        if val in self.getCol(col):
            return False

        # 检查所在的3x3宫格
        if val in self.getBox(row,col):
            return False

        return True

    def inferenceSudoku(self) -> bool:
        """
        使用回溯法推理
        :return: True->推理成功
                 False->推理失败
        """
        for row in range(Grid.GRID_SIZE):
            for col in range(Grid.GRID_SIZE):
                # 0处才需要推理
                val = self.grid[row][col]
                if val != 0:
                    continue

                # 在[row][col]位置枚举1-9所有的可能
                for k in range(1, 10):
                    if self.isValid(row, col, k):
                        self.grid[row][col] = k

                        # 递归调用
                        if self.inferenceSudoku():
                            return True

                        # 回溯
                        self.grid[row][col] = 0

                return False  # 1-9均不满足要求,数独推理失败

        return True  # 没有返回False,数独推理成功

    def clone(self):
        """
        克隆当前数独对象
        :return: 新的Sudoku对象
        """
        newSudoku = Sudoku(self.sudokuStr)
        newSudoku.grid = copy.deepcopy(self.grid)
        return newSudoku

    def serialize(self) -> str:
        """
        将当前数独状态串行化为字符串(便于传输)
        :return: 数独字符串
        """
        return ''.join(str(val) for row in self.grid for val in row)

    def display(self):
        """
        打印在控制台
        """
        for each in self.grid:
            print(each)

    def equal(self, other):
        """
        比较两个Sudoku对象是否相等（比较内容）
        :param other: 另一个Sudoku对象
        :return: bool
        """
        if not isinstance(other, Sudoku):
            return False
        return self.grid == other.grid

File: Exp1/Sudoku-Java/Sudoku.java
package com.sudoku;

import java.util.Arrays;

// 定义一个表示数独网格的类
class Grid {
    // 常量定义
    public static final int GRID_SIZE = 9; // 数独网格的大小
    public static final int BOX_SIZE = 3;   // 小九宫格的大小
    protected int[][] grid;                 // 存储数独的二维数组

    // 构造函数，初始化空的数独网格
    public Grid() {
        this.grid = new int[GRID_SIZE][GRID_SIZE];
    }

    // 获取指定行的值
    public int[] getRow(int row) {
        return grid[row];
    }

    // 获取指定列的值
    public int[] getCol(int col) {
        int[] column = new int[GRID_SIZE];
        for (int i = 0; i < GRID_SIZE; i++) {
            column[i] = grid[i][col];
        }
        return column;
    }

    // 获取指定小九宫格的值
    public int[] getBox(int row, int col) {
        int[] box = new int[BOX_SIZE * BOX_SIZE];
        int rowStart = (row / BOX_SIZE) * BOX_SIZE; // 计算小九宫格的起始行
        int colStart = (col / BOX_SIZE) * BOX_SIZE; // 计算小九宫格的起始列
        int index = 0;
        // 遍历小九宫格并存储值
        for (int i = rowStart; i < rowStart + BOX_SIZE; i++) {
            for (int j = colStart; j < colStart + BOX_SIZE; j++) {
                box[index++] = grid[i][j];
            }
        }
        return box;
    }
}

// 定义一个数独类，继承自 Grid 类
public class Sudoku extends Grid {
    private String sudokuStr; // 存储数独字符串表示

    // 构造函数，接收数独字符串并解析
    public Sudoku(String sudokuStr) {
        super(); // 调用父类构造函数
        this.sudokuStr = sudokuStr; // 设置数独字符串
        parseStr(); // 解析字符串
    }

    // 检查数独字符串的合法性
    public boolean checkStr() {
        return sudokuStr != null && sudokuStr.matches("^[0-9]{81}$");
    }

    // 解析数独字符串并填充网格
    public boolean parseStr() {
        if (!checkStr()) {
            return false; // 字符串不合法
        }
        // 将字符串的每个字符转换为整数并存储到网格中
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                grid[row][col] = Character.getNumericValue(sudokuStr.charAt(row * GRID_SIZE + col));
            }
        }
        return true; // 解析成功
    }

    // 检查在指定位置放置数值的合法性
    public boolean isValid(int row, int col, int val) {
        // 检查行是否存在相同的数值
        if (Arrays.stream(getRow(row)).anyMatch(v -> v == val)) return false;
        // 检查列是否存在相同的数值
        if (Arrays.stream(getCol(col)).anyMatch(v -> v == val)) return false;

        // 检查小九宫格是否存在相同的数值
        int rowStart = (row / BOX_SIZE) * BOX_SIZE;
        int colStart = (col / BOX_SIZE) * BOX_SIZE;
        for (int i = rowStart; i < rowStart + BOX_SIZE; i++) {
            for (int j = colStart; j < colStart + BOX_SIZE; j++) {
                if (grid[i][j] == val) return false;
            }
        }
        return true; // 合法
    }

    // 推理解决数独，采用回溯算法
    public boolean inferenceSudoku() {
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                if (grid[row][col] != 0) continue; // 如果当前格子已填则跳过

                for (int k = 1; k <= 9; k++) {
                    if (isValid(row, col, k)) {
                        grid[row][col] = k; // 填入数值

                        if (inferenceSudoku()) {
                            return true; // 如果成功解决，则返回
                        }

                        grid[row][col] = 0; // 回溯
                    }
                }
                return false; // 如果没有找到合法数值，则返回 false
            }
        }
        return true; // 数独已解决
    }

    // 克隆当前数独对象
    public Sudoku clone() {
        Sudoku newSudoku = new Sudoku(this.sudokuStr); // 创建新的数独对象
        for (int i = 0; i < GRID_SIZE; i++) {
            newSudoku.grid[i] = Arrays.copyOf(this.grid[i], GRID_SIZE); // 复制网格
        }
        return newSudoku;
    }

    // 序列化数独对象为字符串
    public String serialize() {
        StringBuilder sb = new StringBuilder();
        for (int[] row : grid) {
            for (int val : row) {
                sb.append(val); // 将每个数值添加到字符串
            }
        }
        return sb.toString();
    }

    // 打印当前数独网格
    public void display() {
        for (int[] row : grid) {
            System.out.println(Arrays.toString(row)); // 打印每一行
        }
    }

    // 比较两个数独对象是否相等:比较内容，不是比较地址(==比较地址)
    public boolean equal(Sudoku other) {
        return other != null && Arrays.deepEquals(this.grid, other.grid); // 比较网格
    }
}
